Add staking contract

happyleow opened the pull request.

LilaRest commented on contracts/hardhat/deploy/05-LDY.cts:
You need to set the required confirmations to 1 when deploying locally using this pattern, as seen in other deployment scripts
```
import { type DeployFunction } from "hardhat-deploy/dist/types";

module.exports = (async ({ getNamedAccounts, deployments, getChainId }) => {
  const chainId = await getChainId();

  await deployments.deploy("...", {
    from: deployer,
    log: true,
    waitConfirmations: chainId == "31337" ? 1 : 2,
  });
}) as DeployFunction;
```

happyleow replied:
It's resolved.

LilaRest reviewed (changes requested):
Hey Leo!
Thanks for this one
I attached some first comments

Also, the `cc:coverage` command looks broken on my side, which prevents me from testing the coverage of your unit tests.

Here is the error:
```
genhtml: ERROR: line 139 of contracts/src/LDYStaking.sol has branchcov but no linecov data
        (use "genhtml --ignore-errors category ..." to bypass this error)
```

Moreover, can you address all the outputs of `bun cc:slither:report` by else solving the problem, else ignoring the alert with `cc:slither:triage` if not relevant.

Then, can you ensure that you followed all the practices mentioned in this documentation before I go about reviewing the contract?
https://consensys.github.io/smart-contract-best-practices/

Finally, do you mind improving the macro documentation of the staking contract? By explaining the different mechanisms involved, and sending in this PR links to the contracts you took inspiration from (if any).

Thanks!

LilaRest commented on contracts/src/LDYStaking.sol:
Also, that be awesome if you make this contract relying on the GlobalOwner, GlobalPause and GlobalBlacklist. So it can be managed consistently with all other contracts of the protocol.

happyleow replied:
Yea, i agree with you, will modify the contract accordingly.

happyleow commented:
> Hey Leo! Thanks for this one I attached some first comments
> 
> Also, the `cc:coverage` command looks broken on my side, which prevents me from testing the coverage of your unit tests.
> 
> Here is the error:
> 
> ```
> genhtml: ERROR: line 139 of contracts/src/LDYStaking.sol has branchcov but no linecov data
>         (use "genhtml --ignore-errors category ..." to bypass this error)
> ```
> 
> Moreover, can you address all the outputs of `bun cc:slither:report` by else solving the problem, else ignoring the alert with `cc:slither:triage` if not relevant.
> 
> Then, can you ensure that you followed all the practices mentioned in this documentation before I go about reviewing the contract? https://consensys.github.io/smart-contract-best-practices/
> 
> Finally, do you mind improving the macro documentation of the staking contract? By explaining the different mechanisms involved, and sending in this PR links to the contracts you took inspiration from (if any).
> 
> Thanks!

@LilaRest Hello Lila, 
I modified the staking contract and added Natspec formats as you mentioned.
There's only one thing left, I am not able to execute `bun cc:coverage`, it still fails on my side, but I ensure 100% code coverage.
Regarding the staking mechanism, I was referring to the Synthetix staking contract.
Here's the original contract link: https://github.com/Synthetixio/synthetix/blob/develop/contracts/StakingRewards.sol
Please let me know if you need any info.

LilaRest commented:
Hello @happyleow,
Looks cleaner! I'm gonna merge the PR for now so you can move forward with the UI and I can prepare the $LDY contract for CCIP. But we'll still have to thoroughly review everything before deploying the staking contracts then.
Cheers!
